builtin/credential/cert/path_login.go

```diff
@@ -273,10 +291,15 @@ func (b *backend) verifyCredentials(ctx context.Context, req *logical.Request, d
 		for _, tCert := range trust.Certificates { // For each certificate in the entry
 			for _, chain := range trustedChains { // For each root chain that we matched
 				for _, cCert := range chain { // For each cert in the matched chain
-					if tCert.Equal(cCert) && // ParsedCert intersects with matched chain
-						b.matchesConstraints(clientCert, chain, trust) { // validate client cert + matched chain against the config
-						// Add the match to the list
-						matches = append(matches, trust)
+					if tCert.Equal(cCert) { // ParsedCert intersects with matched chain
+						match, err := b.matchesConstraints(ctx, clientCert, chain, trust, verifyConf) // validate client cert + matched chain against the config
```

Comment:
I'm not _quite_ sure this is correct down in the OCSP code, but its easier to explain here.

`tCert` comes via `trusted` off of the storage backend.
`cCert` comes via `trustedChains` off of the `PeerCertificates`: https://github.com/hashicorp/vault/blob/7e50dd0f293308d1cc3f67bfacc3386cd85c7a97/builtin/credential/cert/path_login.go#L653-L680 via `x509.Verify(...)`.

Per [docs](https://pkg.go.dev/crypto/x509#Certificate.Verify), `trustedChains` includes the leaf cert here, which shouldn't matter as we've hopefully excluded non-CA certs from chains before storage... But is in TLS order (`leaf->[ICA->...]root`)...

*However*, the call to `checkForCertInOCSP` is incorrect (and in particular, its `[len(x) - 1]` is incorrect!). In particular, we're using the _last_ cert in the chain, which isn't the leaf's issuer, but could be some arbitrary parent issuer.

This _should_ fail verification, as the OCSP responder doesn't necessarily know about the ultimate Root CA this leaf is issued under, and may only know about the ICA that actually issued this leaf.

Hence I believe down there, given that `trust` includes the leaf in spot 0, the code should use `chain[1]`.

Reply:
Ah, great analysis.  This was code I was very unsure about.